addons/mail/models/discuss/ir_websocket.py

```diff
@@ -82,25 +88,81 @@ def _build_bus_channel_list(self, channels):
         channels.extend([*all_user_channels, *member_specific_channels])
         return super()._build_bus_channel_list(channels)
 
+    def _prepare_subscribe_data(self, channels, last):
+        data = super()._prepare_subscribe_data(channels, last)
+        str_presence_channels = {
+            c for c in channels if isinstance(c, str) and c.startswith("odoo-presence-")
+        }
+        presence_channels = self._build_presence_channel_list(
+            [tuple(c.replace("odoo-presence-", "").split("_")) for c in str_presence_channels]
+        )
+        # There is a gap between a subscription client side (which is debounced)
+        # and the actual subcription thus presences can be missed. Send a
+        # notification to avoid missing presences during a subscription.
+        partners = self.env["res.partner"].browse(
+            [p.id for p, _ in presence_channels if isinstance(p, self.pool["res.partner"])]
+        )
+        guest_ids = [g.id for g, _ in presence_channels if isinstance(g, self.pool["mail.guest"])]
+        # sudo: res.partner - can acess users of partner channels to find
+        # their presences as those channels were already verified during
+        # `_build_bus_channel_list`.
+        domain = expression.AND(
+            [
+                [("last_poll", ">", datetime.now() - timedelta(seconds=2))],
+                expression.OR(
+                    [
+                        [("user_id", "in", partners.with_context(active_test=False).sudo().user_ids.ids)],
+                        [("guest_id", "in", guest_ids)]
+                    ]
+                ),
+            ]
+        )
+        # sudo: mail.presence: can access presences linked to presence channels.
+        missed_presences = self.env["mail.presence"].sudo().search(domain)
+        all_channels = OrderedSet(presence_channels)
+        all_channels.update(
+            self._build_bus_channel_list([c for c in channels if c not in str_presence_channels])
+        )
+        data.update({"channels": all_channels, "missed_presences": missed_presences})
+        return data
+
+    def _subscribe(self, og_data):
+        super()._subscribe(og_data)
+        data = self._prepare_subscribe_data(og_data["channels"], og_data["last"])
+        if bus_target := self._get_missed_presences_bus_target():
```

Comment:
We could potentially remove the presence logic from `prepare_subscribe_data`. Conceptually, presence channels are similar to other channels, so it makes sense to build the list within `_build_bus_channel_list` (we can keep a helper function for clarity). Afterwards, we could call `_dispatch_missed_presences(channels)`, which would handle this specific logic without bloating the data returned by `_prepare_subscribe_data`. This approach better separates concerns: `_prepare_subscribe_data` always returns the same dictionary, regardless of which modules are installed, while the mail module calls `_dispatch_missed_presence` independently, performing only its specific task.